x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authz/store/NativePrivilegeStore.java

```diff
@@ -133,17 +223,69 @@ public void getPrivileges(Collection<String> applications, Collection<String> na
                         .setFetchSource(true)
                         .request();
                     logger.trace(() ->
-                        new ParameterizedMessage("Searching for privileges [{}] with query [{}]", names, Strings.toString(query)));
+                        new ParameterizedMessage("Searching for [{}] privileges with query [{}]",
+                            applications, Strings.toString(query)));
                     request.indicesOptions().ignoreUnavailable();
+                    // TODO: not parsing source of cached entries?
                     ScrollHelper.fetchAllByEntity(client, request, new ContextPreservingActionListener<>(supplier, listener),
                         hit -> buildPrivilege(hit.getId(), hit.getSourceRef()));
                 }
             });
         }
     }
 
-    private boolean isSinglePrivilegeMatch(Collection<String> applications, Collection<String> names) {
-        return applications != null && applications.size() == 1 && hasWildcard(applications) == false && names != null && names.size() == 1;
+    public void invalidate(Collection<String> updatedApplicationNames) {
+        logger.debug("Invalidating application privileges caches for: {}", updatedApplicationNames);
+        numInvalidation.incrementAndGet();
+        // Always completely invalidate application names cache due to wildcard
+        applicationNamesCache.invalidateAll();
+        final Set<String> uniqueNames = Set.copyOf(updatedApplicationNames);
```

Comment:
For invalidation, the application names cache is always completey invalidated since it requires some effort to identify the applicable entries. We could do this, but the gain may not be much.

Reply:
Remind me, what's Kibana's typical usage pattern for querying? Does it use wildcards for the application name?

If so, I think invalidating the name cache means that invalidating for a single application (which might not even exist) would effectively invalidate the whole cache because it would mean that querying for `kibana*` would end up not using any cache at all.

Reply:
Kibana always sends a single concrete application name `kibana-.kibana`. So it should be fine for Kibana's typical usage. 

But let me know if you think it is still necessary. The logic would look like something as the follows:
```java
foreach cache key (type is Set<String>)
   foreach key member
     foreach application
        if (key member == application) or (key member is a wildcard and matches application)
          invalidate the cache key
```

Java code would be
```java
StreamSupport.stream(applicationNamesCache.keys().spliterator(), false)
    .filter(keys -> keys.contains("*")
        || Sets.intersection(keys, uniqueNames).isEmpty() == false
        || keys.stream().filter(k -> k.endsWith("*")).anyMatch(
            k -> uniqueNames.stream().anyMatch(n -> n.regionMatches(false, 0, k, 0, k.length()-1))))
    .forEach(applicationNamesCache::invalidate);

Reply:
Let's leave it - realistically we're talking about Kibana only, so clearing all applications isn't actually going to hurt anyone, and the only time _we_ will clear the cache is on privilege update which happens when you install a new Kibana version.